csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll

```diff
@@ -185,38 +185,23 @@ private class Overflowable extends UnaryOperation {
   }
 }
 
-private class SystemType extends ValueOrRefType {
-  SystemType() { this.getNamespace() instanceof SystemNamespace }
-}
-
-private class SystemTypeMention extends TypeMention {
-  SystemTypeMention() { this.getType() instanceof SystemType }
-
-  Assembly getSystemAssembly() { result = this.getType().getALocation() }
-
-  Assembly getAssembly() { result = this.getTarget().(ControlFlowElement).getAssembly() }
+private class CoreLib extends Assembly {
+  CoreLib() { this = any(SystemExceptionClass c).getALocation() }
 }
 
 /**
  * Holds if assembly `a` was definitely compiled with core library `core`.
- *
- * The analysis is conservative, as it requires mentioning of a (non-special)
- * core type inside assembly `a`.
  */
 pragma[noinline]
-private predicate assemblyCompiledWithCoreLib(Assembly a, Assembly core) {
-  exists(SystemTypeMention stm | a = stm.getAssembly() |
-    core = stm.getSystemAssembly() and
-    // Special built-in types like `object` and `int` are collapsed into one entity
-    // in the presence of multiple core libraries, so such entities cannot be used
-    // to determine the actual core library used at compilation
-    strictcount(stm.getSystemAssembly()) = 1
-  )
+private predicate assemblyCompiledWithCoreLib(Assembly a, CoreLib core) {
+  a.getAnAttribute().getType().getABaseType*().getALocation() = core
```

Comment:
```
core = a.getAnAttribute().getType().getBaseClass*().(SystemAttributeClass).getALocation()
```
otherwise we'll end up getting the location of `ObjectType` which will be all of them.

Also note `getBaseClass()` as we don't care about interfaces.

Reply:
I just remembered that only source assemblies have attributes (due to extractor limitations). I don't think this affects anything her but is worth bearing in mind.

Reply:
Again, good spot :-)